### PR TITLE
refactor: split the steps into separate files, add tests

### PR DESCRIPTION
**Pull request by RyanGWU82**
## 🧰 Changes

This may be my favorite PR I've written since I got to ReadMe. 😁

The "upsert controller" was a long complicated piece of code that ran through a bunch of steps depending on various input parameters and prerequisite conditions. I've rewritten it to loop through a list of "steps" (classes) and run each one in order. Well, it really loops through the steps twice: first to check each step's prereqs and determine whether that step should be run; then a second loop to run the eligible steps.

And the controller is now testable! 🎉 Every class has very straightforward entrypoints that are now testable. The only thing that's not tested is `controllers/upsert/index.js` but that's essentially just these 2 loops and then some pretty output.

I want to do this to Deploybert too, because those controllers are unwieldy and getting pretty unmaintainable… not urgently but maybe during/after the next round of Deploybert updates.

Caveats and notes:
1. Because each step is written independently, this makes a bunch of duplicate Heroku API calls to `heroku.getApp()` and `heroku.appExists()`. I resolved this by memoizing `heroku.getApp()` and `heroku.appExists()` in the heroku module, but that means there's a little more "bookkeeping" to clear the memoized values when appropriate.
2. No programmatic way to say "Be _certain_ we're running the HerokuDeploy step or the DockerDeploy step, but not both." 🤷🏼
3. For testability, I load more parameters in `main.js`'s `getParams()` rather than trying to read them in `heroku.js` or `git.js`.

## 🧬 QA & Testing

I didn't test every single permutation of parameters — this refactor changes a ton of things but most of it seems pretty safe because of the thorough unit tests. Here are the tests I did run:

* [x] [Example 1](https://github.com/readmeio/metrics-test-server/runs/7905887408?check_suite_focus=true): A successful Herkou build, without the `docker` param
* [x] [Example 2](https://github.com/readmeio/metrics-test-server/runs/7905891569?check_suite_focus=true): A successful Heroku build, with `docker` set to `false`
* [x] [Example 3](https://github.com/readmeio/metrics-test-server/runs/7905894431?check_suite_focus=true): A successful Docker build, with `docker` set to `true`
* [x] [Example 4](https://github.com/readmeio/metrics-test-server/runs/7905895077?check_suite_focus=true): A broken Heroku build — required dependency does not exist
* [x] [Example 5](https://github.com/readmeio/metrics-test-server/runs/7905895991?check_suite_focus=true): A broken Docker build

### CODE REVIEW

File: src/controllers/upsert/config-vars.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+const core = require('@actions/core');
+
+const heroku = require('../../heroku');
+
+class ConfigVarsStep {
```
Comment (erunion) [inline src/controllers/upsert/config-vars.js:5]: Might require a move to TS but it'd be cool to have these extend an interface that requires `checkPrereqs` and `run`.

File: test/controllers/upsert/config-vars.test.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+const SAMPLE_APP_NAME = 'owlzilla';
+const SAMPLE_PIPELINE_NAME = 'owlline';
+const SAMPLE_CONFIG_VARS = { VAR1: 'value1', VAR2: 'value2' };
+
+describe('#src/controllers/upsert/config-vars', () => {
+  setupBeforeAndAfter();
+
+  describe('checkPrereqs()', () => {
+    it('should run when the pipeline has 1 or more config vars', async () => {
+      heroku.getPipelineId = jest.fn(() => ({ id: SAMPLE_PIPELINE_ID }));
```
Comment (erunion) [inline test/controllers/upsert/config-vars.test.js:17]: I'm surprised you don't need to mock out `src/heroku` for these overrides to work inside the `config-vars` file.

File: test/heroku.test.js [hunk 3]
```diff
@@ -36,7 +47,21 @@ describe('#src/heroku', () => {
...
+        await expect(heroku.getAppFeature(SAMPLE_APP_NAME, SAMPLE_FEATURE)).resolves.toStrictEqual(SAMPLE_RESPONSE);
+      });
+
+      it('should throw an error if the given app or feature does not exist', async () => {
+        nock('https://api.heroku.com').get(`/apps/${SAMPLE_APP_NAME}/features/${SAMPLE_FEATURE}`).reply(404);
+        await expect(heroku.getAppFeature(SAMPLE_APP_NAME, SAMPLE_FEATURE)).rejects.toThrow(/404/);
+      });
+    });
+
+    describe.skip('getPipelineId()', () => {
```
Comment (erunion) [inline test/heroku.test.js:64]: Some skipped tests in here FYI.

**Review by erunion — APPROVED** [review]
nice! would definitely love to see a rewrite like this in deploybert at some point -- im sure it'd make that codebase easier for more folks to pop in and out of.

File: test/heroku.test.js (hunk 3 above)
Comment (RyanGWU82) [reply test/heroku.test.js]: Thanks, fixed!

File: src/controllers/upsert/config-vars.js (hunk 1 above)
Comment (RyanGWU82) [reply src/controllers/upsert/config-vars.js]: Great call — done! See https://github.com/readmeio/heroku-review-app-action/pull/35/commits/e8bf4032e49d6c2823543598a19a51d0c803c9a9

**Pull request merged**